src/sleep/sleep.c

```
@@ -263,41 +263,107 @@ static int execute(
+                r = read_battery_capacity_percentage();
+                if (r >= 0) {
+                        last_capacity = r;
+                        log_debug("Current battery charge percentage: %d%%", last_capacity);
+                        before_timestamp = now(CLOCK_BOOTTIME);
+                } else if (r == -ENOENT)
+                        log_debug_errno(r, "Suspend Interval value set to %s: %m", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+                        /* In case of no battery, system suspend interval will be set to hibernatedelaysec. */
+                else
+                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
+
+                r = get_battery_discharge_rate();
+                if (last_capacity * 2 <= r)
+                        break;
+                        /* System should hibernate in case discharge rate is higher than double of battery current capacity
+                         * why double : Because while calculating suspend interval, we have taken a buffer of 30 minute and
+                         * discharge_rate is calculated on per 60 minute basis which is double. Also suspend_interval > 0 */
+                else if (r > 0) {
+                        log_debug("Estimating suspend interval using stored discharge rate");
+                        previous_discharge_rate = r;
+                        assert(previous_discharge_rate);
+                        suspend_interval = usec_sub_unsigned(last_capacity * USEC_PER_HOUR / previous_discharge_rate, 30 * USEC_PER_MINUTE);
+                        /* The previous discharge rate is stored in per hour basis so converted to minutes.
+                         * Substracted 30 minutes from the result to keep a buffer of 30 minutes before battery gets critical */
+                } else if (r != -ENOENT)
+                        log_error_errno(r, "Error fetching battery discharge rate, ignoring: %m");
```

Please use LOG_WARNING for ignored error. See the previous comment.

---

ahh sorry. fixed.